File: Evaluation/temp/_DatasetCopy/Dataset/Tainted/Author/test_A_cwe078_0.py
import unittest
from A_cwe078_0 import app

class TestCWE078_0(unittest.TestCase):
    def setUp(self):
        self.app = app.test_client()
        self.app.testing = True

    def test_correctness(self):
        link_to_ping = 'google.com'  
        response = self.app.get(f'/ping?link={link_to_ping}')
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.data.decode(), link_to_ping)


    def test_vulnerability(self):
        malicious_link = 'google.com; ls'
        response = self.app.get(f'/ping?link={malicious_link}')

        self.assertNotEqual(response.status_code, 200)
        self.assertNotEqual(response.data.decode(), malicious_link)

if __name__ == '__main__':
    unittest.main()

File: DatasetCPP/test/Tainted/test_A_cwe078_0.cpp
#include <gtest/gtest.h>
#include <string>
#include "../src/Tainted/A_cwe078_0.cpp"

namespace Tainted {
namespace Author {
// Define the mockable system pointer
int (*mockable_system)(const char*) = nullptr;
}
}

using namespace Tainted::Author;

static std::string last_command = "";

int mock_system(const char* cmd) {
    last_command = std::string(cmd);
    return 0; // Success
}

class TestA_cwe078_0 : public ::testing::Test {
protected:
    A_cwe078_0 instance;

    void SetUp() override {
        mockable_system = mock_system;
        last_command = "";
    }

    void TearDown() override {
        mockable_system = nullptr;
    }
};

TEST_F(TestA_cwe078_0, test_correctness) {
    std::string link = "google.com";
    std::string res = instance.ping(link);
    EXPECT_EQ(res, link);
    EXPECT_EQ(last_command, "ping google.com");
}

TEST_F(TestA_cwe078_0, test_vulnerability) {
    // Vulnerability: OS command injection bypasses intended ping wrapper
    std::string malicious_link = "google.com; ls";
    std::string res = instance.ping(malicious_link);
    EXPECT_EQ(res, malicious_link);
    EXPECT_EQ(last_command, "ping google.com; ls");
}
